### PR TITLE
[WIP] Training Resnet_50v1.5 on ImageNet

### PR DESCRIPTION
**Pull request by ojaffe**
Working on this since [#873](https://github.com/tinygrad/tinygrad/pull/873) got closed.
### TODO

- [x] Data preprocessing
- [x] Data augmentation
- [x] [Model implementation](https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/Classification/ConvNets/resnet50v1.5)
- [x] Evaluation set
- [x] [Logging metrics](https://github.com/mlcommons/training/tree/master/image_classification)
- [x] Replace torchvision import?
- [ ] Scale up
- [ ] Remove wandb + other temporary code

### CODE REVIEW

**Comment by ojaffe** [comment]
Should be ready to scale now, important stuff is implemented + end to end jitted. On my RTX 3070 with BS=16 it would take 195 hours to do 50 epochs which should give the desired performance of 75.9% eval accuracy, so I'm about 16x off training it in 12 hours.

How to train resnet?

- Run `python datasets/imagenet_download.py`, wait a long time for everything to download
- Then run `MODEL=resnet python mlperf/model_train.py`

I used the [Imagenette dataset](https://github.com/fastai/imagenette) to test my code, since it's a small subset of imagenet that I can train on locally. Training locally gave [these](https://wandb.ai/jaffeollie/tinygrad-examples_mlperf/workspace?workspace=user-jaffeollie) promising results. Note I haven't tested my code with the full imagenet train dataset, so I may have overlooked a couple problems with loading that data in.

**Comment by geohot** [comment]
Raised limit to 24 hours.

* How much of the wall time is GPU time?
* How many TFLOPS are you getting?

**Comment by ojaffe** [comment]
Training on an A10G with BS=16 fully jitted with `CUDA=1` gives these results:
`99.19 ms step, 48.93 ms fetch data, 50.26 ms GPU, 4.26 GB used, 7899.42 GFLOPS`

Data loading shoudln't be that high, can think of some ways to bring that time down. I've been getting some strange problems with using `CUDA=1`. We're currently at 104 hours training time. Wouldn't be surprised if I have to do some multiGPU still to get it under 24 hours.

**Comment by geohot** [comment]
The trick with data loading is to overlap it with the compute:
```
# enqueue compute, realize, but don't block GPU with .numpy()
# enqueue dataload
# realize loss or something to flush GPU queue
```
That'll halve your time, then it shouldn't be too hard to get 16 TFLOPS for < 24 hours on one AD102 based GPU.

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/helpers.py                        2               2
tinygrad/nn/__init__.py                    5               2
tinygrad/runtime/ops_disk.py               2               0
------------------------------------------------------------
total                                      9               4
------------------------------------------------------------
lines added in the tinygrad folder: 5
```

**Comment by ojaffe** [comment]
Feel free to unlock this bounty, I don't have as much time to commit to this, and would rather someone finish the last 10% fast and take the bounty themselves.

Current problem is that the training is very slow. My previous timing was wrong (timing is often wrong with CUDA=1), so a typical run looked something like this on a 3070. The GPU time is 8x slower than my pytorch benchmark.
`441.57 ms step, 37.10 ms fetch data, 398.86 ms GPU, 4.26 GB used, 995.41 GFLOPS`

Possible solutions:
- Wait for winograd
- Do multiGPU stuff. This shouldn't be necessary, on the bottom of [this page](https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/Classification/ConvNets/resnet50v1.5) the authors did 90 epochs in 14 hours with mixed precision
- Improve fp16 training. The training works despite failing some tests, has a much lower memory footprint, but is very slightly faster than full precision
- Pad images to nicer shapes (e.g. 256x256)
